File: examples/cpp/selection_operator.cpp
#include <cstddef>
#include <cstdint>
#include <vector>

#include <tsl.hpp>

struct negative_op {
    template <class Vec>
    typename Vec::mask_type operator()(typename tsl::reg_param<Vec>::type value) const {
        return tsl::less_than<Vec>(
            value,
            tsl::set1<Vec>(static_cast<typename Vec::base_type>(0)));
    }
};

struct less_than_op {
    template <class Vec>
    typename Vec::mask_type operator()(
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) const {
        return tsl::less_than<Vec>(left, right);
    }
};

void fill_input(std::vector<std::int32_t>& input) {
    for (std::size_t i = 0; i < input.size(); ++i) {
        input[i] = static_cast<std::int32_t>(static_cast<int>((i * 17) % 53) - 26);
    }
}

void fill_inputs(
    std::vector<std::int32_t>& left,
    std::vector<std::int32_t>& right) {
    for (std::size_t i = 0; i < left.size(); ++i) {
        left[i] = static_cast<std::int32_t>(static_cast<int>((i * 17) % 53) - 26);
        right[i] = static_cast<std::int32_t>(static_cast<int>((i * 11) % 47) - 23);
    }
}

template <class Parallelism>
bool run_unary_selection_case() {
    constexpr std::size_t count = 1000;
    constexpr std::int32_t sentinel = 1234567;
    std::vector<std::int32_t> input(count);
    std::vector<std::int32_t> output(count, sentinel);
    fill_input(input);

    const auto produced = tsl::algo::select_unary<
        Parallelism,
        tsl::algo::alignment::unaligned>(
        negative_op{},
        input.data(),
        output.data(),
        count);

    std::size_t expected_count = 0;
    for (std::size_t i = 0; i < count; ++i) {
        if (input[i] < 0) {
            if (output[expected_count] != input[i]) {
                return false;
            }
            expected_count += 1;
        }
    }
    if (produced != expected_count) {
        return false;
    }
    for (std::size_t i = produced; i < count; ++i) {
        if (output[i] != sentinel) {
            return false;
        }
    }
    return true;
}

template <class Parallelism>
bool run_binary_selection_case() {
    constexpr std::size_t count = 1003;
    constexpr std::int32_t sentinel = 7654321;
    std::vector<std::int32_t> left(count);
    std::vector<std::int32_t> right(count);
    std::vector<std::int32_t> output(count, sentinel);
    fill_inputs(left, right);

    const auto produced = tsl::algo::select_binary<
        Parallelism,
        tsl::algo::alignment::unaligned>(
        less_than_op{},
        left.data(),
        right.data(),
        output.data(),
        count);

    std::size_t expected_count = 0;
    for (std::size_t i = 0; i < count; ++i) {
        if (left[i] < right[i]) {
            if (output[expected_count] != left[i]) {
                return false;
            }
            expected_count += 1;
        }
    }
    if (produced != expected_count) {
        return false;
    }
    for (std::size_t i = produced; i < count; ++i) {
        if (output[i] != sentinel) {
            return false;
        }
    }

    output.assign(count, sentinel);
    const auto range_produced = tsl::algo::select_binary<Parallelism>(
        less_than_op{},
        left,
        right,
        output);
    if (range_produced != expected_count) {
        return false;
    }
    expected_count = 0;
    for (std::size_t i = 0; i < count; ++i) {
        if (left[i] < right[i]) {
            if (output[expected_count] != left[i]) {
                return false;
            }
            expected_count += 1;
        }
    }
    for (std::size_t i = expected_count; i < count; ++i) {
        if (output[i] != sentinel) {
            return false;
        }
    }
    return true;
}

template <class Parallelism>
bool run_selection_cases() {
    return run_unary_selection_case<Parallelism>() &&
           run_binary_selection_case<Parallelism>();
}

int main() {
    if (!run_selection_cases<tsl::dataparallel::fixed<1>>()) {
        return 1;
    }
    if (!run_selection_cases<tsl::dataparallel::generic<4>>()) {
        return 2;
    }
    if (!run_selection_cases<tsl::dataparallel::generic<16>>()) {
        return 3;
    }
    return 0;
}

File: examples/rust/src/bin/selection_operator.rs
use tsl_generated::tsl_core::StaticSimdVector;
use tsl_generated::tsl_scalar as tsl;

struct Negative;

impl<V> tsl::algo::UnaryPredicateKernel<V> for Negative
where
    V: StaticSimdVector<BaseType = i32>
        + tsl::detail::primitives::Less_thanImpl
        + tsl::detail::primitives::Set1Impl,
{
    fn test(&mut self, value: V::RegisterType) -> V::MaskType {
        tsl::less_than::<V>(value, tsl::set1::<V>(0))
    }
}

struct LessThan;

impl<V> tsl::algo::BinaryPredicateKernel<V> for LessThan
where
    V: StaticSimdVector<BaseType = i32> + tsl::detail::primitives::Less_thanImpl,
{
    fn test(&mut self, left: V::RegisterType, right: V::RegisterType) -> V::MaskType {
        tsl::less_than::<V>(left, right)
    }
}

fn fill_input(input: &mut [i32]) {
    for (i, value) in input.iter_mut().enumerate() {
        *value = ((i * 17) % 53) as i32 - 26;
    }
}

fn fill_inputs(left: &mut [i32], right: &mut [i32]) {
    for (i, (left_value, right_value)) in left.iter_mut().zip(right.iter_mut()).enumerate() {
        *left_value = ((i * 17) % 53) as i32 - 26;
        *right_value = ((i * 11) % 47) as i32 - 23;
    }
}

fn verify_selected<F, G>(
    output: &[i32],
    produced: usize,
    count: usize,
    mut predicate: F,
    mut selected_value: G,
) where
    F: FnMut(usize) -> bool,
    G: FnMut(usize) -> i32,
{
    let mut expected_count = 0usize;
    for i in 0..count {
        if predicate(i) {
            assert_eq!(output[expected_count], selected_value(i));
            expected_count += 1;
        }
    }
    assert_eq!(produced, expected_count);
    for value in output.iter().take(count).skip(produced) {
        assert_eq!(*value, i32::MAX);
    }
}

fn main() {
    macro_rules! run_policy {
        ($policy:expr) => {{
            let policy = $policy;

            let mut input = vec![0i32; 1000];
            fill_input(&mut input);
            let mut output = vec![i32::MAX; input.len()];
            let mut negative = Negative;
            let produced = tsl::algo::select_unary(policy, &mut negative, &input, &mut output);
            verify_selected(
                &output,
                produced,
                input.len(),
                |i| input[i] < 0,
                |i| input[i],
            );

            let mut left = vec![0i32; 1003];
            let mut right = vec![0i32; 1003];
            fill_inputs(&mut left, &mut right);
            output.resize(left.len(), i32::MAX);
            output.fill(i32::MAX);
            let mut less_than = LessThan;
            let produced =
                tsl::algo::select_binary(policy, &mut less_than, &left, &right, &mut output);
            verify_selected(
                &output,
                produced,
                left.len(),
                |i| left[i] < right[i],
                |i| left[i],
            );
        }};
    }

    run_policy!(tsl::algo::parallelism::native());
    run_policy!(tsl::algo::parallelism::fixed::<1>());
    run_policy!(tsl::algo::parallelism::generic::<4>());
    run_policy!(tsl::algo::parallelism::generic::<16>());
}
